### PR TITLE
dqn model

### PR DESCRIPTION
**Pull request by bnb32**
Added DQN model for training. Takes environment as input. Basic usage example included.

### CODE REVIEW

File: examples/run_linear_solver.py [hunk 1]
```diff
@@ -1,9 +1,14 @@
 """Run linear solver"""
```
Comment (Khev) [inline examples/run_linear_solver.py]: """Run linear solver""" ---> """Run linear solver: solve a x + b = 0"""

File: examples/run_linear_solver.py [hunk 2]
```diff
@@ -1,9 +1,14 @@
...
+from rl_equation_solver.environment.algebraic import AlgebraicEnv
+from rl_equation_solver.agent.agent import Agent
+from sympy import symbols
 
 if __name__ == '__main__':
     init_logger(__name__, log_level='DEBUG')
     init_logger('rl_equation_solver', log_level='DEBUG')
 
-    Env.run()
+    env = AlgebraicEnv(symbols('x a b'))
```
Comment (Khev) [inline examples/run_linear_solver.py]: ideally we'd change:

AlgebraicEnv(symbols('x a b'))

To

AlgebraicEnv(order=2)

here and have order define the symbols

File: rl_equation_solver/agent/agent.py [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+                                     lr=Config.LR, amsgrad=True)
+
+    @property
+    def history(self):
+        """Get training history"""
+        return self.env.history
+
+    def choose_optimal_action(self, state):
+        """
+        Choose action with max reward based on given state
```
Comment (Khev) [inline rl_equation_solver/agent/agent.py]: Chose action with max "exptected reward := max a Q(s,a)"

File: rl_equation_solver/agent/agent.py [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+        """
+        with torch.no_grad():
+            return self.policy_network(state).max(1)[1].view(1, 1)
+
+    def choose_action(self, state):
+        """
+        Choose action based on given state. Either choose optimal action or
+        random action depending on training step.
+        """
+        sample = random.random()
```
Comment (Khev) [inline rl_equation_solver/agent/agent.py]: Rename "sample" --> "random_float" for clarity. My first thought was that "sample" was an (S,A,R,S') sample from the replay buffer

File: rl_equation_solver/agent/agent.py [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+        decay = (Config.EPSILON_START - Config.EPSILON_END)
+        decay *= math.exp(-1. * self.steps_done / Config.EPSILON_DECAY)
+        epsilon_threshold = Config.EPSILON_END + decay
+
+        self.steps_done += 1
+        if sample > epsilon_threshold:
+            return self.choose_optimal_action(state)
+        else:
+            return torch.tensor([[self.env.action_space.sample()]],
```
Comment (Khev) [inline rl_equation_solver/agent/agent.py]: Suggestion: defined a variable 'random_action' = torch.tensor....  and then return that, for greater interpretability

File: rl_equation_solver/agent/agent.py [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+                                f"state = {self.env.state_string}")
+                    break
+
+                losses = self.env.history['loss'][-Config.RESET_STEPS:]
+                if len(losses) >= Config.RESET_STEPS and len(set(losses)) <= 1:
+                    logger.info(f'Loss has been constant ({list(losses)[0]}) '
+                                f'for {Config.RESET_STEPS} steps. Reseting.')
+                    break
+
+    def predict(self, state_string):
```
Comment (Khev) [inline rl_equation_solver/agent/agent.py:281]: Future: maybe add a epsilon optional argument to allow for (eps-greedy) stochastic policy

File: rl_equation_solver/environment/algebraic.py [hunk 7]
```diff
@@ -119,64 +129,67 @@ def _get_symbols(self):
...
+            String representation of the current state
 
-        Example
+        Returns
         -------
-        keys = ['Add', 'Mul', 'Pow'] + ['x', 'a', 'b']
-        return {key: -(i + 2) for i, key in enumerate(keys)}
-        """
+        loss : int
+            Number of edges plus number of nodes in graph representation of
```
Comment (Khev) [inline rl_equation_solver/environment/algebraic.py]: "graph representation" --> "graph representation / expression tree" .... some people use the "expression tree: lingo

File: rl_equation_solver/environment/algebraic.py [hunk 8]
```diff
@@ -250,36 +266,53 @@ def step(self, action: int):
...
+
+    def update_history(self, entry):
+        """Update training history of policy_network"""
+        self._history['loss'].append(entry['loss'])
+        self._history['reward'].append(entry['reward'])
+        self._history['state'].append(entry['state'])
 
     def find_reward(self, state_old, state_new):
         """
         Reward is decrease in loss
```
Comment (Khev) [inline rl_equation_solver/environment/algebraic.py:294]: "Reward is decrease in loss" --> "Reward is (stepwise) decrease in loss"

File: rl_equation_solver/environment/algebraic.py [hunk 9]
```diff
@@ -250,36 +266,53 @@ def step(self, action: int):
...
+        reward : int
+            Difference between loss for state_new and state_old
         """
         loss_old = self.find_loss(state_old)
         loss_new = self.find_loss(state_new)
         return loss_old - loss_new
 
+    def _get_feature_dict(self):
+        """Return feature dict representing features at each node"""
+        keys = ['Add', 'Mul', 'Pow']
```
Comment (Khev) [inline rl_equation_solver/environment/algebraic.py]: Do we use this method currently? If so, why do we exclude ['Sub','Div'] from the keys? IIRC its because sympy doesn't use those in the expression tree; we need to check this out

**Review by Khev — APPROVED** [review]
lgtm

File: rl_equation_solver/agent/agent.py (hunk 6 above)
Comment (Khev) [reply rl_equation_solver/agent/agent.py:281]: Or maybe the stochasticity is already contained in the choose optimal action?

File: rl_equation_solver/environment/algebraic.py (hunk 9 above)
Comment (bnb32) [reply rl_equation_solver/environment/algebraic.py]: We do use it, in the `to_vec` method, and I had the same question. I dont think those operators are excluded from the expression tree though. I was also confused about the indexing here - why `-(i+2)`?

File: rl_equation_solver/environment/algebraic.py (hunk 8 above)
Comment (bnb32) [reply rl_equation_solver/environment/algebraic.py:294]: I'm gonna keep this as is because we could try out non integer reward functions.

File: examples/run_linear_solver.py (hunk 2 above)
Comment (bnb32) [reply examples/run_linear_solver.py]: Agreed. Good call.

File: rl_equation_solver/agent/agent.py (hunk 6 above)
Comment (bnb32) [reply rl_equation_solver/agent/agent.py:281]: This is included in the training routine with the `choose_action` method, which uses either `choose_optimal_action` or `choose_random_action` based on epsilon.
